src/hashset.c

```diff
@@ -0,0 +1,1681 @@
+
+            /* Emit elements in the smaller table, if this bucket hasn't already
+             * been rehashed. */
+            if (table_small == 0 && !cursorIsLessThan(cursor, s->rehashIdx)) {
+                bucket *b = &s->tables[table_small][cursor & mask_small];
+                if (b->presence) {
+                    for (int pos = 0; pos < ELEMENTS_PER_BUCKET; pos++) {
+                        if (b->presence & (1 << pos)) {
+                            void *emit = emit_ref ? &b->elements[pos] : b->elements[pos];
+                            fn(privdata, emit);
+                        }
+                    }
+                }
+                in_probe_sequence |= b->everfull;
+            }
+
+            /* Iterate over indices in larger table that are the expansion of
+             * the index pointed to by the cursor in the smaller table. */
+            do {
+                /* Emit elements in bigger table. */
+                bucket *b = &s->tables[table_large][cursor & mask_large];
+                if (b->presence) {
+                    for (int pos = 0; pos < ELEMENTS_PER_BUCKET; pos++) {
+                        if (b->presence & (1 << pos)) {
+                            void *emit = emit_ref ? &b->elements[pos] : b->elements[pos];
+                            fn(privdata, emit);
+                        }
+                    }
+                }
+                in_probe_sequence |= b->everfull;
+
+                /* Increment the reverse cursor not covered by the smaller mask.*/
+                cursor = nextCursor(cursor, mask_large);
+
+                /* Continue while bits covered by mask difference is non-zero */
+            } while (cursor & (mask_small ^ mask_large) && cursor != start_cursor);
+        }
+        if (cursor == 0) {
+            cursor_passed_zero = 1;
+        }
+    } while (in_probe_sequence && !single_step && cursor != start_cursor);
+    hashsetResumeRehashing(s);
+    return cursor_passed_zero ? 0 : cursor;
+}
+
+/* --- Iterator --- */
+
+/* Initiaize a iterator, that is not allowed to insert, delete or even lookup
+ * elements in the hashset, because such operations can trigger incremental
+ * rehashing which moves elements around and confuses the iterator. Only
+ * hashsetNext is allowed. Each element is returned exactly once. Call
+ * hashsetResetIterator when you are done. See also hashsetInitSafeIterator. */
+void hashsetInitIterator(hashsetIterator *iter, hashset *s) {
+    iter->hashset = s;
+    iter->table = 0;
+    iter->index = -1;
+    iter->safe = 0;
+}
+
+/* Initialize a safe iterator, which is allowed to modify the hash table while
+ * iterating. It pauses incremental rehashing to prevent elements from moving
+ * around. Call hashsetNext to fetch each element. You must call
+ * hashsetResetIterator when you are done with a safe iterator.
+ *
+ * Guarantees:
+ *
+ * - Elements that are in the hash table for the entire iteration are returned
+ *   exactly once.
```

Comment:
Can't an element be returned twice from the first bucket if we wrap around zero? I feel like you could guard against this to make this guarantee.

Reply:
The iterator doesn't wrap around zero.

Are you thinking about scan?

Scan can return an element twice if it follows a probing chain wrapping around zero. I don't know how to prevent that, but it seems to be allowed according to the scan docs.

Reply:
hashsetNext doesn't follow probe chains though, and doesn't wrap around. Rehashing is paused and not moving anything, and hashsetReplace won't move an item to another bucket either.

Reply:
Yeah, I was confusing this with scan.

I guess I also noticed that you aren't support to modify a hashset when safe iteration is enabled. There is an assertion on the fingerprint that includes stuff like number of elements. So the later guarantees in this comment are misuses of the iterator.

Reply:
Yes. Exactly like in dict.